### PR TITLE
only Tensors of floating point dtype can require gradients (see #7021)

### PR DESCRIPTION
**Pull request by t-vi**
Sometimes, people are surprised that things cannot be differentiated w.r.t. integer parameters such indices.
The following patch takes some steps to prevent them from requiring gradients of non-floating point Tensors:
- by setting `tensor.requires_grad = True`
- by using `tensor.requires_grad_(True)`
- by using factory functions with `requires_grad=True`

Of course, applying the above with `False` needs to still be allowed.

This as requested by Adam in #7021, this is done at the Python interface level.

### CODE REVIEW

**Comment by t-vi** [comment]
something is up with the pytorch-linux-xenial-py3-clang5-asan test. It seems to hang for 30 minutes in test_multi_drop (test_utils.TestDataLoader).

**Comment by zou3519** [comment]
@t-vi Probably just the test, I've seen intermittent timeouts there

**Comment by t-vi** [comment]
Indeed. Worked when I changed the commit msg. There is an issue about it, too.

**Review by ezyang — APPROVED** [review]

File: test/test_autograd.py [hunk 1]
```diff
@@ -2157,6 +2157,42 @@ def test_diagonal_derivative_requires_grad(self):
...
+            def f3():
+                torch.ones(1, dtype=dt, device='cuda' if cuda else 'cpu', requires_grad=True)
+            a = torch.ones(1, dtype=dt, device='cuda' if cuda else 'cpu')
+            a.requires_grad = False  # should always work
+            a.requires_grad_(False)
+
+            for f in [f1, f2, f3]:
+                a = torch.ones(1, dtype=dt, device='cuda' if cuda else 'cpu')
+                if dt.is_floating_point:
+                    f()
```
Comment (apaszke) [inline test/test_autograd.py]: Can you make those functions take `a` as an argument? I'd say that depending on the scoping rules of Python (that they see `a` that's defined at every loop iteration) is very confusing.

**Comment by t-vi** [comment]
So, now the macos has CI changed, but I think works.

**Comment by ezyang** [comment]
OS X failure is unrelated, and I think fixed on master. @pytorchbot retest this please

**Review by apaszke — APPROVED** [review]

**Comment by gchanan** [comment]
This didn't change the constructors in `tensor_new.cpp`, e.g. `torch.tensor`.

If you implemented this in those constructors, it would get a little awkward when combined with type inference, because you don't know the type of the tensor that will come out, e.g.:

```
def convert_to_tensors(data0, data1)
  return torch.tensor(data0, requires_grad=True), torch.tensor(data1, requires_grad=True)
```

would not throw an error on `convert_to_tensors([0., 1.], [2., 3.]) but would on convert_to_tensors([0., 1.], [2, 3]).  Sometimes you want this fail-fast behavior, but sometimes not.
